scripts/2.TCGA-MLexample-covariates.py

```diff
@@ -80,108 +79,112 @@
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
 
-X_partitions = {
-    'full': {
-        'train': X.ix[train_index], 
-        'test': X.ix[test_index]
-        },
-    'expressions': {
-        'train': expression_pca.ix[train_index], 
-        'test': expression_pca.ix[test_index]
-        },
-    'covariates': {
-        'train': covariates.ix[train_index], 
-        'test': covariates.ix[test_index]
-        }    
-    } 
-
-y_train = y[train_index]
-y_test = y[test_index]
-
-'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(
-    len(X_partitions['full']['train'].columns), 
-    len(X_partitions['full']['train']), 
-    len(X_partitions['full']['test']))
+# Select the feature set for the different models
+N_COVARIATES = covariates.shape[1]
+def select_feature_set_columns(X, feature_set):
+    if feature_set=='expressions': return X[:, N_COVARIATES:]
+    return  X[:, :N_COVARIATES]
 
 
 # ## Define pipeline and Cross validation model fitting
 
-# In[14]:
+# In[11]:
 
 # Parameter Sweep for Hyperparameters
-param_grid = {
-    'classify__loss': ['log'],
-    'classify__penalty': ['elasticnet'],
-    'classify__alpha': [10 ** x for x in range(-3, 1)],
-    'classify__l1_ratio': [0, 0.2, 0.8, 1],
+n_components_list = [50, 100]
+
+param_grids = {
+    'full': {
+        'features__expressions__pca__n_components' : n_components_list,
+        'classify__loss': ['log'],
+        'classify__penalty': ['elasticnet'],
+        'classify__alpha': [10 ** x for x in range(-3, 1)],
+        'classify__l1_ratio': [0, 0.2, 0.8, 1],
+    },
+    'expressions': {
+        'features__expressions__pca__n_components' : n_components_list,
+        'classify__loss': ['log'],
+        'classify__penalty': ['elasticnet'],
+        'classify__alpha': [10 ** x for x in range(-3, 1)],
+        'classify__l1_ratio': [0, 0.2, 0.8, 1],
+    },
+    'covariates': {
+        'classify__loss': ['log'],
+        'classify__penalty': ['elasticnet'],
+        'classify__alpha': [10 ** x for x in range(-3, 1)],
+        'classify__l1_ratio': [0, 0.2, 0.8, 1],
+    }
 }
 
-pipeline = Pipeline(steps=[
+# Feature selection
+expression_features = Pipeline([
+    ('select_features', FunctionTransformer(select_feature_set_columns,
+        kw_args={'feature_set': 'expressions'})),
     ('standardize', StandardScaler()),
-    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+    ('pca', PCA())
+])
+covariate_features = Pipeline([
+    ('select_features', FunctionTransformer(select_feature_set_columns,
+        kw_args={'feature_set': 'covariates'})),
+    ('standardize', StandardScaler())
 ])
 
+# Full model pipelines
+pipeline_definitions = {
+    'full': Pipeline([
+        ('features', FeatureUnion([
+            ('expressions', expression_features),
+            ('covariates', covariate_features)
+        ])),
+        ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
```

Comment:
Let's define these once above:

```python
sgd_classifier_params = {
    'random_state': 0,
    'class_weight': 'balanced',
}

... SGDClassifier(**sgd_classifier_params)
```

Could even put the invariant `loss` and `penalty` here (unless there is some `sklearn` bug.